### ISSUE TITLE
Custom integer type serialization TypeError

### ISSUE THREAD

**Issue by mikeshultz** [pydantic/pydantic#10152]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

I ran into an unusual serialization issue with custom `int` types. Seems to only occur if it's is included in a dict with `Any` type hint. And only with on integers greater than some threshold (maybe greater than 32 byte). For instance, `9e18` will not cause the issue, but `10e18` will raise.

Running the provided test script will raise a `TypeError` during `model_dum(mode="json")`.

```
Traceback (most recent call last):
  File "/home/mike/dev/test-scripts/pydantic_error.py", line 33, in <module>
    ).model_dump(mode="json")
      ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mike/.venvs/test-scripts/lib/python3.12/site-packages/pydantic/main.py", line 364, in model_dump
    return self.__pydantic_serializer__.to_python(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: expected int, got 10000000000000000000
```

I've only been able to repro the issue under the following conditions:

- a custom `int` type is involved
- the `int` value is above a certain threshold
- the custom type is a value in a dict within another model
- the type hint for the dict does not include the custom type (e.g. `Any`)
- `model_dump()` must be called with the kwarg `mode="json"`

I've included a minimal test case. The `__get_pydantic_core_schema__` classmethod is not required to produce the error. Let me know if I can provide any other information. 

Thanks!

### Example Code

```Python
from typing import Any

from pydantic import BaseModel
from pydantic_core.core_schema import (
    CoreSchema,
    int_schema,
    plain_serializer_function_ser_schema,
    with_info_before_validator_function,
)


class KindaInt(int):
    # This is only necessary for the union type in Container
    @classmethod
    def __get_pydantic_core_schema__(cls, value, handler=None) -> CoreSchema:
        schema = with_info_before_validator_function(lambda x, y: int(x), int_schema())
        schema["serialization"] = plain_serializer_function_ser_schema(
            function=lambda x: int(x)
        )
        return schema


class Container(BaseModel):
    # Works
    # stuff: dict[str, KindaInt | int | str | float | None]
    # Does not work
    stuff: dict[str, Any]


print(
    Container(
        # Works
        # stuff={"value": KindaInt(int(1e18))},
        # Does not work
        stuff={"value": KindaInt(int(10e18))},
    ).model_dump(mode="json")
)
```

### Python, Pydantic & OS Version

```Text
/home/mike/.venvs/test-scripts/lib/python3.12/site-packages/pydantic/_migration.py:283: UserWarning: `pydantic.utils:version_info` has been moved to `pydantic.version:version_info`.
  warnings.warn(f'`{import_path}` has been moved to `{new_location}`.')
             pydantic version: 2.8.2
        pydantic-core version: 2.20.1
          pydantic-core build: profile=release pgo=true
                 install path: /home/mike/.venvs/test-scripts/lib/python3.12/site-packages/pydantic
               python version: 3.12.4 (main, Jun  7 2024, 06:33:07) [GCC 14.1.1 20240522]
                     platform: Linux-6.10.3-arch1-2-x86_64-with-glibc2.40
             related packages: mypy-1.8.0 fastapi-0.110.3 pydantic-settings-2.2.1 typing_extensions-4.11.0
                       commit: unknown
```

**Comment by sydney-runkle** [comment]
Hmm, my guess would be a challenge with `BigInt` inference in `pydantic-core` for custom types... perhaps a good first issue for someone interested in getting to know `pydantic-core` a bit better. Thanks for the bug report!

**Comment by kxx317** [comment]
The bug occurs when the value exceeds the i64 integer bound. 9e18 is smaller than that while 10e18 is larger, resulting in an OverFlowError when extracting the value into an i64 value (in pyo3).

I have an open pull request for a fix in `pydantic-core` that addresses this issue.

**Comment by sydney-runkle** [comment]
It's possible this will be released in a patch (v2.9.1), as we might not get to another `pydantic-core` version before v2.9...

**Comment by mikeshultz** [comment]
Appreciate the quick response on this. Thank you.
